### PR TITLE
feat: send correct x-goog-api-client headers for the diregapic

### PR DESCRIPTION
**Pull request by viacheslav-rostovtsev**
This makes the `x-goog-api-client` header for the diregapic libraries to have a `rest` component and sends the headers with the request.

### CODE REVIEW

File: Google.Api.Gax.Grpc/Rest/RestChannel.cs [hunk 1]
```diff
@@ -64,10 +66,24 @@ public RestChannel(RestServiceCollection serviceCollection, string endpoint, Cha
...
         private async Task<HttpResponseMessage> SendAsync<TRequest>(RestMethod restMethod, string host, CallOptions options, TRequest request, CancellationToken cancellationToken)
         {
-            // TODO: Add headers from options.Headers, but intercept the x-goog-api-client header.
+            var headers = options.Headers
+                .Where(mh => !mh.IsBinary)
+                .Where(mh=> mh.Key != VersionHeaderBuilder.HeaderName)
+                .Select(mh => new KeyValuePair<string, string>(mh.Key, mh.Value))
+                .ToList();
+
+            var restVersion = new VersionHeaderBuilder()
```
Comment (jskeet) [inline Google.Api.Gax.Grpc/Rest/RestChannel.cs:75]: We could presumably construct this once and store it as a static variable, right? I think that would be better, as I don't know what the efficiency of these methods is like.

File: Google.Api.Gax.Grpc/Rest/RestChannel.cs [hunk 2]
```diff
@@ -64,10 +66,24 @@ public RestChannel(RestServiceCollection serviceCollection, string endpoint, Cha
...
+                .AppendAssemblyVersion("gax", typeof(CallSettings))
+                .AppendAssemblyVersion("rest", typeof(HttpClient))
+                .ToString();
+
+            headers.Add(new KeyValuePair<string, string>(VersionHeaderBuilder.HeaderName, restVersion));
+
             // Ideally, add the header in the client builder instead of in the ServiceSettingsBase...
             // TODO: Use options. How do we set the timeout for an individual HTTP request? We probably need to create a linked cancellation token.
-            var httpRequest = restMethod.CreateRequest((IMessage) request, host);
+            var httpRequest = restMethod.CreateRequest((IMessage) request, host, headers);
```
Comment (jskeet) [inline Google.Api.Gax.Grpc/Rest/RestChannel.cs:86]: I'm not sure that I'd put the header addition into RestMethod, given that nothing about the headers depends on the method. If we keep the header addition in this class instead, we don't need to materialize the list - just add all text headers (other than the version) from the options, then add the version header separately.

File: Google.Api.Gax/VersionHeaderBuilder.cs [hunk 3]
```diff
@@ -62,6 +62,12 @@ public VersionHeaderBuilder AppendAssemblyVersion(string name, System.Type type)
         /// Appends the .NET environment information to the list.
         /// </summary>
         public VersionHeaderBuilder AppendDotNetEnvironment() => AppendVersion("gl-dotnet", s_environmentVersion.Value);
+        
+        /// <summary>
+        /// Whether the name or value that are supposed to be included in a header are valid
+        /// </summary>
+        private static bool IsHeaderNameValueValid(string nameOrValue)
+            => !nameOrValue.Contains(" ") && !nameOrValue.Contains("/");
```
Comment (jskeet) [inline Google.Api.Gax/VersionHeaderBuilder.cs:70]: Nit: we normally put the `=>` at the end of the previous line.

File: Google.Api.Gax/VersionHeaderBuilder.cs [hunk 4]
```diff
@@ -103,16 +109,15 @@ private static string GetEntryAssemblyVersionOrNull()
...
+        
         private static string FormatAssemblyVersion(System.Type type)
         {
             // Prefer AssemblyInformationalVersion, then AssemblyFileVersion,
             // then AssemblyVersion.
 
             var assembly = type.GetTypeInfo().Assembly;
             var info = assembly.GetCustomAttributes<AssemblyInformationalVersionAttribute>().FirstOrDefault()?.InformationalVersion;
-            if (info != null)
+            if (info != null && IsHeaderNameValueValid(info)) // Skip informational version if it's not a valid header value
```
Comment (jskeet) [inline Google.Api.Gax/VersionHeaderBuilder.cs:120]: I assume this is because you've seen this going wrong somewhere? Out of interest, what happened?

Comment (viacheslav-rostovtsev) [reply Google.Api.Gax/VersionHeaderBuilder.cs:120]: The informational version for `System.Net.Http` is `4.6.30015.01 @BuiltBy: AzDevOps-vseng20170000ZL @Branch: release/2.1-MSRC @SrcCode: https://github.com/dotnet/corefx/tree/96d2fdd8610e6eead9b357789266be9d149dd0b8`
So the code would take it and immediately throw when checking it for spaces and slashes.

File: Google.Api.Gax.Grpc/Rest/RestChannel.cs (hunk 1 above)
Comment (viacheslav-rostovtsev) [reply Google.Api.Gax.Grpc/Rest/RestChannel.cs:75]: 👍

File: Google.Api.Gax.Grpc/Rest/RestChannel.cs (hunk 2 above)
Comment (viacheslav-rostovtsev) [reply Google.Api.Gax.Grpc/Rest/RestChannel.cs:86]: I've moved it but it feels strange to me that there is a method named `CreateRequest` and it creates a request object … except for the headers.

File: Google.Api.Gax.Grpc/Rest/RestMethod.cs [hunk 5]
```diff
@@ -143,13 +144,15 @@ internal HttpRequestMessage CreateRequest(IMessage protoRequest, string host)
             var uri = host is null ? new Uri(uriPathWithParams, UriKind.Relative) : new UriBuilder { Host = host, Path = uriPathWithParams }.Uri;
             
             var content = transcodingResult.Body is null ? null : new StringContent(transcodingResult.Body, Encoding.UTF8, s_applicationJsonMediaType);
-            
-            return new HttpRequestMessage
+
+            var httpRequestMessage = new HttpRequestMessage
```
Comment (jskeet) [inline Google.Api.Gax.Grpc/Rest/RestMethod.cs:148]: Revert all the changes to this file now?

File: Google.Api.Gax/VersionHeaderBuilder.cs (hunk 4 above)
Comment (jskeet) [reply Google.Api.Gax/VersionHeaderBuilder.cs:120]: Ick. That's horrible.

**Review by jskeet — APPROVED** [review]
While I see your point about creating a semi-incomplete request, I'd say that RestMethod is creating a request with all the information that comes from the rest method itself... other code can contribute other aspects. (There could even be further sources of information eventually, I guess.)

I definitely like it better now - I'd just suggest resetting RestMethod itself back to the previous version.

File: Google.Api.Gax/VersionHeaderBuilder.cs (hunk 3 above)
Comment (viacheslav-rostovtsev) [reply Google.Api.Gax/VersionHeaderBuilder.cs]: :+1:

File: Google.Api.Gax.Grpc/Rest/RestMethod.cs (hunk 5 above)
Comment (viacheslav-rostovtsev) [reply Google.Api.Gax.Grpc/Rest/RestMethod.cs]: :+1:

**Pull request merged**